From ba414ae811f381dab82bd0c63bc5dc50f5fda90c Mon Sep 17 00:00:00 2001
From: Marcin Wojtas <mw@semihalf.com>
Date: Wed, 24 Jul 2019 07:57:59 +0200
Subject: [PATCH 352/386] telephony: mvebu_phone: enable low-level IRQ
 configuration via TAL

Current SLIC drivers wish to control low-level interrupt configuration
of the TDM controller. Replace a mechanism of calling mv_phone_dev_
routines directly with according TAL callbacks. It allows to stop
relying on global variables and split TDM2C / TDMMC handling properly.
Remove unused TDMMC routines.

Change-Id: I37fe5e589373cbf3be56ef6ef45736c44073b777
Signed-off-by: Marcin Wojtas <mw@semihalf.com>
Reviewed-on: https://sj1git1.cavium.com/12960
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Stefan Chulski <stefanc@marvell.com>
[RH: Original patch taken from marvell 88F3720 board support SDK 10.0-PR2003]
Signed-off-by: Ruiqiang Hao <Ruiqiang.Hao@windriver.com>
---
 drivers/telephony/mvebu_phone/mv_phone_dev.c  | 46 ++++++-------------
 .../telephony/mvebu_phone/slic/drv_dxt_if.c   |  5 +-
 .../telephony/mvebu_phone/slic/silabs_if.c    |  5 +-
 .../telephony/mvebu_phone/slic/zarlink_if.c   |  5 +-
 drivers/telephony/mvebu_phone/tal/tal.c       | 14 ++++++
 drivers/telephony/mvebu_phone/tal/tal.h       |  4 ++
 drivers/telephony/mvebu_phone/tdmmc/tdmmc.c   |  8 ----
 7 files changed, 41 insertions(+), 46 deletions(-)

diff --git a/drivers/telephony/mvebu_phone/mv_phone_dev.c b/drivers/telephony/mvebu_phone/mv_phone_dev.c
index c33e7cb53224..f7ee63c333d2 100644
--- a/drivers/telephony/mvebu_phone/mv_phone_dev.c
+++ b/drivers/telephony/mvebu_phone/mv_phone_dev.c
@@ -710,6 +710,18 @@ static void tdm_if_stats_get(struct tal_stats *tdm_if_stats)
 		tdm2c_ext_stats_get(&tdm_if_stats->tdm_ext_stats);
 }
 
+/* Enable device interrupts. */
+void tdm2c_if_intr_enable(void)
+{
+	tdm2c_intr_enable();
+}
+
+/* Disable device interrupts. */
+void tdm2c_if_intr_disable(void)
+{
+	tdm2c_intr_disable();
+}
+
 static struct tal_if tdm2c_if = {
 	.pcm_start	= tdm2c_if_pcm_start,
 	.pcm_stop	= tdm2c_if_pcm_stop,
@@ -718,6 +730,8 @@ static struct tal_if tdm2c_if = {
 	.control	= tdm_if_control,
 	.write		= tdm2c_if_write,
 	.stats_get	= tdm_if_stats_get,
+	.intr_enable	= tdm2c_if_intr_enable,
+	.intr_disable	= tdm2c_if_intr_disable,
 };
 
 static struct tal_if tdmmc_if = {
@@ -732,38 +746,6 @@ static struct tal_if tdmmc_if = {
 
 /* Additional helper routines */
 
-/* Enable device interrupts. */
-void mv_phone_intr_enable(u8 dev_id)
-{
-	switch (priv->tdm_type) {
-	case MV_TDM_UNIT_TDM2C:
-		tdm2c_intr_enable();
-		break;
-	case MV_TDM_UNIT_TDMMC:
-		tdmmc_intr_enable(dev_id);
-		break;
-	default:
-		dev_err(&priv->parent->dev, "%s: undefined TDM type\n",
-			__func__);
-	}
-}
-
-/* Disable device interrupts. */
-void mv_phone_intr_disable(u8 dev_id)
-{
-	switch (priv->tdm_type) {
-	case MV_TDM_UNIT_TDM2C:
-		tdm2c_intr_disable();
-		break;
-	case MV_TDM_UNIT_TDMMC:
-		tdmmc_intr_disable(dev_id);
-		break;
-	default:
-		dev_err(&priv->parent->dev, "%s: undefined TDM type\n",
-			__func__);
-	}
-}
-
 /* Get board type for SLIC unit (pre-defined). */
 u32 mv_phone_get_slic_board_type(void)
 {
diff --git a/drivers/telephony/mvebu_phone/slic/drv_dxt_if.c b/drivers/telephony/mvebu_phone/slic/drv_dxt_if.c
index eac36ed64fe5..8dc1d63e29d7 100644
--- a/drivers/telephony/mvebu_phone/slic/drv_dxt_if.c
+++ b/drivers/telephony/mvebu_phone/slic/drv_dxt_if.c
@@ -5,6 +5,7 @@
  *
  */
 
+#include <tal/tal.h>
 #include "drv_dxt_if.h"
 
 static int drv_dxt_spi_cs;
@@ -23,14 +24,14 @@ void drv_dxt_if_signal_interrupt(void)
 void drv_dxt_if_enable_irq(u32 irq)
 {
 	/* We have only one TDM channel */
-	mv_phone_intr_enable(0);
+	tal_intr_enable();
 }
 EXPORT_SYMBOL(drv_dxt_if_enable_irq);
 
 void drv_dxt_if_disable_irq(u32 irq)
 {
 	/* We have only one TDM channel */
-	mv_phone_intr_disable(0);
+	tal_intr_disable();
 }
 EXPORT_SYMBOL(drv_dxt_if_disable_irq);
 
diff --git a/drivers/telephony/mvebu_phone/slic/silabs_if.c b/drivers/telephony/mvebu_phone/slic/silabs_if.c
index 01be96f64fa8..04dcae35c25c 100644
--- a/drivers/telephony/mvebu_phone/slic/silabs_if.c
+++ b/drivers/telephony/mvebu_phone/slic/silabs_if.c
@@ -5,19 +5,20 @@
  *
  */
 
+#include <tal/tal.h>
 #include "silabs_if.h"
 
 static void (*isi_handler)(unsigned long);
 
 void silabs_if_enable_irq(u32 device)
 {
-	mv_phone_intr_enable(device);
+	tal_intr_enable();
 }
 EXPORT_SYMBOL(silabs_if_enable_irq);
 
 void silabs_if_disable_irq(u32 device)
 {
-	mv_phone_intr_disable(device);
+	tal_intr_disable();
 }
 EXPORT_SYMBOL(silabs_if_disable_irq);
 
diff --git a/drivers/telephony/mvebu_phone/slic/zarlink_if.c b/drivers/telephony/mvebu_phone/slic/zarlink_if.c
index 409f4dc6d3c7..4378ff4deb30 100644
--- a/drivers/telephony/mvebu_phone/slic/zarlink_if.c
+++ b/drivers/telephony/mvebu_phone/slic/zarlink_if.c
@@ -5,19 +5,20 @@
  *
  */
 
+#include <tal/tal.h>
 #include "zarlink_if.h"
 
 static void (*zsi_handler)(unsigned long);
 
 void zarlink_if_enable_irq(u32 device)
 {
-	mv_phone_intr_enable(device);
+	tal_intr_enable();
 }
 EXPORT_SYMBOL(zarlink_if_enable_irq);
 
 void zarlink_if_disable_irq(u32 device)
 {
-	mv_phone_intr_disable(device);
+	tal_intr_disable();
 }
 EXPORT_SYMBOL(zarlink_if_disable_irq);
 
diff --git a/drivers/telephony/mvebu_phone/tal/tal.c b/drivers/telephony/mvebu_phone/tal/tal.c
index 1d8cd922b8ce..f0f3b1ac3cc0 100644
--- a/drivers/telephony/mvebu_phone/tal/tal.c
+++ b/drivers/telephony/mvebu_phone/tal/tal.c
@@ -121,3 +121,17 @@ enum tal_status tal_mmp_tx(u8 *buffer, int size)
 	return TAL_STAT_BAD_PARAM;
 }
 EXPORT_SYMBOL(tal_mmp_tx);
+
+void tal_intr_enable(void)
+{
+	if (tal_if && tal_if->intr_enable)
+		tal_if->intr_enable();
+}
+EXPORT_SYMBOL(tal_intr_enable);
+
+void tal_intr_disable(void)
+{
+	if (tal_if && tal_if->intr_disable)
+		tal_if->intr_disable();
+}
+EXPORT_SYMBOL(tal_intr_disable);
diff --git a/drivers/telephony/mvebu_phone/tal/tal.h b/drivers/telephony/mvebu_phone/tal/tal.h
index fa11081d235b..fadb1980750d 100644
--- a/drivers/telephony/mvebu_phone/tal/tal.h
+++ b/drivers/telephony/mvebu_phone/tal/tal.h
@@ -59,6 +59,8 @@ struct tal_if {
 	int (*control)(int cmd, void *data);
 	int (*write)(u8 *buffer, int size);
 	void (*stats_get)(struct tal_stats *tal_stats);
+	void (*intr_enable)(void);
+	void (*intr_disable)(void);
 };
 
 /* API */
@@ -74,5 +76,7 @@ enum tal_status tal_set_if(struct tal_if *interface);
 enum tal_status tal_mmp_rx(u8 *buffer, int size);
 enum tal_status tal_mmp_tx(u8 *buffer, int size);
 enum tal_status tal_write(u8 *buffer, int size);
+void tal_intr_enable(void);
+void tal_intr_disable(void);
 
 #endif /* _TAL_H */
diff --git a/drivers/telephony/mvebu_phone/tdmmc/tdmmc.c b/drivers/telephony/mvebu_phone/tdmmc/tdmmc.c
index 845b964712ff..17ca95e3d2f4 100644
--- a/drivers/telephony/mvebu_phone/tdmmc/tdmmc.c
+++ b/drivers/telephony/mvebu_phone/tdmmc/tdmmc.c
@@ -289,14 +289,6 @@ static void tdmmc_mcdma_stop(void)
 	udelay(1);
 }
 
-void tdmmc_intr_enable(u8 device_id)
-{
-}
-
-void tdmmc_intr_disable(u8 device_id)
-{
-}
-
 void tdmmc_show(void)
 {
 	u32 index;
-- 
2.17.1

